File: lib/generators/apis_file_generator.dart
import 'dart:io';
import 'dart:convert';
import 'package:path/path.dart' as path;
import 'package:http/http.dart' as http;
import 'package:dio/dio.dart';

import '../helpers/fetch_swagger_data.dart';
import '../type/route_info.dart';

// Function to generate the API file
void generateApiFile({
  required String controllerName,
  required String controllerDir,
  required List<TRouteInfo> routesInfo,
  required String? getControllerFuncFromSettings,
  required String apiNameConverter,
  required String dataFetcher,
  required bool enableConfig,
}) {
  if (dataFetcher == "axios") {
    _generateAxios(
      controllerName: controllerName,
      controllerDir: controllerDir,
      routesInfo: routesInfo,
      getControllerFuncFromSettings: getControllerFuncFromSettings,
      apiNameConverter: apiNameConverter,
      enableConfig: enableConfig,
    );
  } else {
    _generateFetch(
      controllerName: controllerName,
      controllerDir: controllerDir,
      routesInfo: routesInfo,
      getControllerFuncFromSettings: getControllerFuncFromSettings,
      apiNameConverter: apiNameConverter,
      enableConfig: enableConfig,
    );
  }
}

// Function to generate the API file using Axios
void _generateAxios({
  required String controllerName,
  required String controllerDir,
  required List<TRouteInfo> routesInfo,
  required String? getControllerFuncFromSettings,
  required String apiNameConverter,
  required bool enableConfig,
}) {
  final folderName = convertToKebabCase(controllerName);
  final filePath = path.join(
    controllerDir,
    '${convertToKebabCase(controllerName)}.api.dart',
  );

  var content = "import 'package:http/http.dart' as http;\n"
      "import 'package:dio/dio.dart';\n"
      "import '../helpers/fetch_swagger_data.dart';\n"
      "import '${toCamelCase(controllerName)}.api_routes.dart';\n";

  List<String> interfacesToImport = [];
  Map<String, bool> routeFlag = {};

  for (var routeInfo in routesInfo) {
    final route = routeInfo.routes.keys.first;
    final info = routeInfo.routes.values.first;
    var apiName = getApiNameFromRoute(
      route,
      true,
      getControllerFuncFromSettings,
      apiNameConverter,
    );

    if (apiName == null) continue;

    var i = 1;
    while (routeFlag.containsKey(apiName)) {
      apiName = '$apiName${i++}';
    }
    routeFlag[apiName] = true;

    if (info.hasQueryParams || info.hasBodyPayload || info.hasPathParams) {
      interfacesToImport.add('I${apiName}Request');
    }
    if (info.hasResponse) {
      interfacesToImport.add('I${apiName}Response');
    }
  }
  interfacesToImport.sort((a, b) => a.length - b.length);

  routeFlag = {};

  if (interfacesToImport.isNotEmpty) {
    content += "import '${folderName}.interface.dart' as i;\n";
  }

  content +=
      "// ----------------------------------------------------------------------\n\n";

  for (var routeInfo in routesInfo) {
    final route = routeInfo.routes.keys.first;
    final info = routeInfo.routes.values.first;
    var apiName = getApiNameFromRoute(
      route,
      false,
      getControllerFuncFromSettings,
      apiNameConverter,
    );
    var apiNameWithController = getApiNameFromRoute(
      route,
      true,
      getControllerFuncFromSettings,
      apiNameConverter,
    );

    if (apiName == null) continue;
    if (apiNameWithController == null) continue;

    var i = 1;
    while (routeFlag.containsKey(apiName)) {
      apiName = '$apiName${i++}';
    }
    routeFlag[apiName] = true;

    if (apiName != apiNameWithController) {
      var j = 1;
      while (routeFlag.containsKey(apiNameWithController)) {
        apiNameWithController = '$apiNameWithController${j++}';
      }
      routeFlag[apiNameWithController] = true;
    }

    content +=
        'Future<i.${handleResponseInApi(info.hasResponse, apiNameWithController)}> ${toCamelCase(apiName)}(${info.hasPathParams || info.hasQueryParams || info.hasBodyPayload ? 'i.${apiNameWithController}Request payload' : ''}) async {\n'
        '  final dio = Dio();\n'
        '  final response = await dio.${info.methodType.toString().split('.').last}(${handleParameterInPathForApi('\${toCamelCase(controllerName)}ApiRoutes.${apiName}', info.hasPathParams)}, data: payload.bodyPayload, queryParameters: payload.queryParams);\n'
        '  return response.data;\n'
        '}\n\n';
  }

  routeFlag = {};

  content +=
      "// ----------------------------------------------------------------------\n\n";

  content += 'class ${toCamelCase(controllerName)}Api {\n';

  List<String> apisToExport = [];

  for (var routeInfo in routesInfo) {
    var route = routeInfo.routes.keys.first;
    var apiName = getApiNameFromRoute(
      route,
      false,
      getControllerFuncFromSettings,
      apiNameConverter,
    );
    if (apiName == null) continue;

    var i = 1;
    while (routeFlag.containsKey(apiName)) {
      apiName = '$apiName${i++}';
    }
    routeFlag[apiName] = true;

    apisToExport.add(toCamelCase(apiName));
  }

  apisToExport.sort((a, b) => a.length - b.length);

  if (apisToExport.isNotEmpty) {
    content += '  ${apisToExport.join(',\n  ')}\n}\n\n';
  }

  content +=
      "// ----------------------------------------------------------------------\n\n";

  content += 'export default ${toCamelCase(controllerName)}Api;';

  File(filePath).writeAsStringSync(content);
}

// Function to generate the API file using Fetch
void _generateFetch({
  required String controllerName,
  required String controllerDir,
  required List<TRouteInfo> routesInfo,
  required String? getControllerFuncFromSettings,
  required String apiNameConverter,
  required bool enableConfig,
}) {
  final folderName = convertToKebabCase(controllerName);
  final filePath = path.join(
    controllerDir,
    '${convertToKebabCase(controllerName)}.api.dart',
  );

  final withHelpersFile = routesInfo.any((routeInfo) {
    final info = routeInfo.routes.values.first;
    return info.hasQueryParams;
  });

  var content = "import 'package:http/http.dart' as http;\n"
      "import '../helpers/fetch_swagger_data.dart';\n"
      "import '${toCamelCase(controllerName)}.api_routes.dart';\n";

  if (withHelpersFile) {
    content +=
        "import '../helpers/helpers.dart';\n"; // Assuming a helpers.dart file
  }

  List<String> interfacesToImport = [];
  Map<String, bool> routeFlag = {};

  for (var routeInfo in routesInfo) {
    var route = routeInfo.routes.keys.first;
    var info = routeInfo.routes[route]!;
    var apiName = getApiNameFromRoute(
      route,
      true,
      getControllerFuncFromSettings,
      apiNameConverter,
    );
    if (apiName == null) continue;

    var i = 1;
    while (routeFlag.containsKey(apiName)) {
      apiName = '$apiName${i++}';
    }
    routeFlag[apiName] = true;

    if (info.hasQueryParams || info.hasBodyPayload || info.hasPathParams) {
      interfacesToImport.add('I${apiName}Request');
    }
    if (info.hasResponse) {
      interfacesToImport.add('I${apiName}Response');
    }
  }
  interfacesToImport.sort((a, b) => a.length - b.length);
  routeFlag = {};

  if (interfacesToImport.isNotEmpty) {
    content += "import '${folderName}.interface.dart' as i;\n";
  }

  content +=
      "// ----------------------------------------------------------------------\n\n";

  for (var routeInfo in routesInfo) {
    var route = routeInfo.routes.keys.first;
    var info = routeInfo.routes[route]!;

    var apiName = getApiNameFromRoute(
      route,
      false,
      getControllerFuncFromSettings,
      apiNameConverter,
    );
    var apiNameWithController = getApiNameFromRoute(
      route,
      true,
      getControllerFuncFromSettings,
      apiNameConverter,
    );

    if (apiName == null) continue;
    if (apiNameWithController == null) continue;

    var i = 1;
    while (routeFlag.containsKey(apiName)) {
      apiName = '$apiName${i++}';
    }
    routeFlag[apiName] = true;

    if (apiName != apiNameWithController) {
      var j = 1;
      while (routeFlag.containsKey(apiNameWithController)) {
        apiNameWithController = '$apiNameWithController${j++}';
      }
      routeFlag[apiNameWithController] = true;
    }

    content +=
        'Future<i.${handleResponseInApi(info.hasResponse, apiNameWithController)}> ${toCamelCase(apiName)}(${info.hasPathParams || info.hasQueryParams || info.hasBodyPayload ? 'i.${apiNameWithController}Request payload' : ''}) async {\n'
        '  final url = \'${handleParameterInPathForApi('\${toCamelCase(controllerName)}ApiRoutes.${apiName}', info.hasPathParams)}${info.hasQueryParams ? '?${/*prepareQueryParams(payload.queryParams)*/ ''}' : ''}\';\n'
        '  final response = await http.get(Uri.parse(url));\n'
        '  if (response.statusCode == 200) {\n'
        '    return i.${handleResponseInApi(info.hasResponse, apiNameWithController)}.fromJson(jsonDecode(response.body));\n'
        '  } else {\n'
        '    throw Exception(\'Failed to load data\');\n'
        '  }\n'
        '}\n\n';
  }
  routeFlag = {};

  content +=
      "// ----------------------------------------------------------------------\n\n";

  content += 'class ${toCamelCase(controllerName)}Api {\n';

  List<String> apisToExport = [];

  for (var routeInfo in routesInfo) {
    var route = routeInfo.routes.keys.first;
    var apiName = getApiNameFromRoute(
      route,
      false,
      getControllerFuncFromSettings,
      apiNameConverter,
    );
    if (apiName == null) continue;

    var i = 1;
    while (routeFlag.containsKey(apiName)) {
      apiName = '$apiName${i++}';
    }
    routeFlag[apiName] = true;

    apisToExport.add(toCamelCase(apiName));
  }

  apisToExport.sort((a, b) => a.length - b.length);

  if (apisToExport.isNotEmpty) {
    content += '  ${apisToExport.join(',\n  ')}\n}\n\n';
  }

  content +=
      "// ----------------------------------------------------------------------\n\n";

  content += 'export default ${toCamelCase(controllerName)}Api;';

  File(filePath).writeAsStringSync(content);
}

File: typescript/src/generators/apis-file-generator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const fetch_swagger_data_helpers_js_1 = require("../helpers/fetch-swagger-data.helpers.js");
const generateAxios = ({ controllerName, controllerDir, routesInfo, getControllerFuncFromSettings, apiNameConverter, enableConfig, }) => {
    const folderName = (0, fetch_swagger_data_helpers_js_1.convertToKebabCase)(controllerName);
    const filePath = path.join(controllerDir, `${(0, fetch_swagger_data_helpers_js_1.convertToKebabCase)(controllerName)}.api.ts`);
    let content = `import ApiInstance from "@api-instance";${enableConfig ? `\nimport { AxiosRequestConfig } from "axios";` : ""}

import ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}ApiRoutes from "./${folderName}.api-routes";\n`;
    const interfacesToImport = [];
    let routeFlag = {};
    routesInfo.forEach((routeInfo) => {
        const [route, info] = Object.entries(routeInfo)[0];
        let apiName = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, true, getControllerFuncFromSettings, apiNameConverter);
        let i = 1;
        while (routeFlag[apiName]) {
            apiName = `${apiName}${i}`;
            i++;
        }
        routeFlag[apiName] = true;
        if (info.hasQueryParams || info.hasBodyPayload || info.hasPathParams) {
            interfacesToImport.push(`I${apiName}Request`);
        }
        if (info.hasResponse) {
            interfacesToImport.push(`I${apiName}Response`);
        }
    });
    interfacesToImport.sort((a, b) => a.length - b.length);
    routeFlag = {};
    if (interfacesToImport.length) {
        content += `import {\n\t${interfacesToImport.join(",\n\t")}\n} from "./${folderName}.interface";\n`;
    }
    content +=
        "\n// ----------------------------------------------------------------------\n\n";
    routesInfo.forEach((routeInfo) => {
        const [route, info] = Object.entries(routeInfo)[0];
        let apiName = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, false, getControllerFuncFromSettings, apiNameConverter);
        let apiNameWithController = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, true, getControllerFuncFromSettings, apiNameConverter);
        let i = 1;
        while (routeFlag[apiName]) {
            apiName = `${apiName}${i}`;
            i++;
        }
        routeFlag[apiName] = true;
        if (apiName !== apiNameWithController) {
            let j = 1;
            while (routeFlag[apiNameWithController]) {
                apiNameWithController = `${apiNameWithController}${j}`;
                j++;
            }
            routeFlag[apiNameWithController] = true;
        }
        content += `const ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(apiName)} = async (${info.hasPathParams || info.hasQueryParams || info.hasBodyPayload
            ? `payload: I${apiNameWithController}Request${enableConfig ? `, axiosConfig?: AxiosRequestConfig` : ``}`
            : enableConfig
                ? "axiosConfig?: AxiosRequestConfig"
                : ""}) => {
  const { data } = await ApiInstance.${info.methodType}${(0, fetch_swagger_data_helpers_js_1.handleResponseInApi)(info.hasResponse, apiNameWithController)}(\n\t\t${(0, fetch_swagger_data_helpers_js_1.handleParameterInPathForApi)(`${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}ApiRoutes.${apiName}`, info.hasPathParams)},${enableConfig ? `\n\t\t` : ``}${(0, fetch_swagger_data_helpers_js_1.handleParameterInQueryAndBodyForApi)(info.hasQueryParams, info.hasBodyPayload, info.isArrayBodyPayload, info.methodType, enableConfig)}\n\t);
  return data;
};\n\n`;
    });
    routeFlag = {};
    content +=
        "// ----------------------------------------------------------------------\n\n";
    content += `const ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}Api = {\n`;
    const apisToExport = [];
    routesInfo.forEach((routeInfo) => {
        const route = Object.keys(routeInfo)[0];
        let apiName = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, false, getControllerFuncFromSettings, apiNameConverter);
        let i = 1;
        while (routeFlag[apiName]) {
            apiName = `${apiName}${i}`;
            i++;
        }
        routeFlag[apiName] = true;
        apisToExport.push((0, fetch_swagger_data_helpers_js_1.toCamelCase)(apiName));
    });
    apisToExport.sort((a, b) => a.length - b.length);
    if (apisToExport.length) {
        content += `\t${apisToExport.join(",\n\t")}\n}\n\n`;
    }
    content +=
        "// ----------------------------------------------------------------------\n\n";
    content += `export default ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}Api;`;
    fs.writeFileSync(filePath, content);
};
const generateFetch = ({ controllerName, controllerDir, routesInfo, getControllerFuncFromSettings, apiNameConverter, enableConfig, }) => {
    const folderName = (0, fetch_swagger_data_helpers_js_1.convertToKebabCase)(controllerName);
    const filePath = path.join(controllerDir, `${(0, fetch_swagger_data_helpers_js_1.convertToKebabCase)(controllerName)}.api.ts`);
    const withHelpersFile = routesInfo.some((routeInfo) => {
        const [route, info] = Object.entries(routeInfo)[0];
        return info.hasQueryParams;
    });
    let content = `import ApiInstance from "@api-instance";\n\n`;
    if (withHelpersFile) {
        content += `import { prepareQueryParams } from "../helpers";\n`;
    }
    content += `import ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}ApiRoutes from "./${folderName}.api-routes";\n`;
    const interfacesToImport = [];
    let routeFlag = {};
    routesInfo.forEach((routeInfo) => {
        const [route, info] = Object.entries(routeInfo)[0];
        let apiName = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, true, getControllerFuncFromSettings, apiNameConverter);
        let i = 1;
        while (routeFlag[apiName]) {
            apiName = `${apiName}${i}`;
            i++;
        }
        routeFlag[apiName] = true;
        if (info.hasQueryParams || info.hasBodyPayload || info.hasPathParams) {
            interfacesToImport.push(`I${apiName}Request`);
        }
        if (info.hasResponse) {
            interfacesToImport.push(`I${apiName}Response`);
        }
    });
    interfacesToImport.sort((a, b) => a.length - b.length);
    routeFlag = {};
    if (interfacesToImport.length) {
        content += `import {\n\t${interfacesToImport.join(",\n\t")}\n} from "./${folderName}.interface";\n`;
    }
    content +=
        "\n// ----------------------------------------------------------------------\n\n";
    routesInfo.forEach((routeInfo) => {
        const [route, info] = Object.entries(routeInfo)[0];
        let apiName = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, false, getControllerFuncFromSettings, apiNameConverter);
        let apiNameWithController = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, true, getControllerFuncFromSettings, apiNameConverter);
        let i = 1;
        while (routeFlag[apiName]) {
            apiName = `${apiName}${i}`;
            i++;
        }
        routeFlag[apiName] = true;
        if (apiName !== apiNameWithController) {
            let j = 1;
            while (routeFlag[apiNameWithController]) {
                apiNameWithController = `${apiNameWithController}${j}`;
                j++;
            }
            routeFlag[apiNameWithController] = true;
        }
        content += `const ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(apiName)} = async (${info.hasPathParams || info.hasQueryParams || info.hasBodyPayload
            ? `payload: I${apiNameWithController}Request${enableConfig ? `, fetchConfig?: RequestInit` : ""}`
            : enableConfig
                ? "fetchConfig?: RequestInit"
                : ""})${info.hasResponse ? `: Promise<I${apiNameWithController}Response>` : ``} => {\n`;
        content += `  const response = await ApiInstance${info.hasResponse ? `<I${apiNameWithController}Response>` : ""}(${info.hasQueryParams ? "`${" : ""}${(0, fetch_swagger_data_helpers_js_1.handleParameterInPathForApi)(`${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}ApiRoutes.${apiName}`, info.hasPathParams)}${info.hasQueryParams ? "}" : ""}${info.hasQueryParams ? "?${prepareQueryParams(payload.queryParams)}" : ""}${info.hasQueryParams ? "`" : ""}, {
    ${enableConfig ? `...fetchConfig,\n` : ""}method: '${info.methodType}',
    headers: {
      'Content-Type': '${info.mediaType}',${enableConfig ? `\n      ...fetchConfig?.headers,` : ""}
    },${info.hasBodyPayload
            ? `\n    body: JSON.stringify(payload.bodyPayload),`
            : ""}
  });\n`;
        content += `  return response;\n};\n\n`;
    });
    routeFlag = {};
    content +=
        "// ----------------------------------------------------------------------\n\n";
    content += `const ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}Api = {\n`;
    const apisToExport = [];
    routesInfo.forEach((routeInfo) => {
        const route = Object.keys(routeInfo)[0];
        let apiName = (0, fetch_swagger_data_helpers_js_1.getApiNameFromRoute)(route, false, getControllerFuncFromSettings, apiNameConverter);
        let i = 1;
        while (routeFlag[apiName]) {
            apiName = `${apiName}${i}`;
            i++;
        }
        routeFlag[apiName] = true;
        apisToExport.push((0, fetch_swagger_data_helpers_js_1.toCamelCase)(apiName));
    });
    apisToExport.sort((a, b) => a.length - b.length);
    if (apisToExport.length) {
        content += `\t${apisToExport.join(",\n\t")}\n}\n\n`;
    }
    content +=
        "// ----------------------------------------------------------------------\n\n";
    content += `export default ${(0, fetch_swagger_data_helpers_js_1.toCamelCase)(controllerName)}Api;`;
    fs.writeFileSync(filePath, content);
};
const generateApiFile = ({ controllerName, controllerDir, routesInfo, getControllerFuncFromSettings, apiNameConverter, dataFetcher, enableConfig, }) => {
    if (dataFetcher === "axios") {
        generateAxios({
            routesInfo,
            enableConfig,
            controllerDir,
            controllerName,
            apiNameConverter,
            getControllerFuncFromSettings,
        });
    }
    else {
        generateFetch({
            routesInfo,
            enableConfig,
            controllerDir,
            controllerName,
            apiNameConverter,
            getControllerFuncFromSettings,
        });
    }
};
exports.default = generateApiFile;
